repondeur/zam_repondeur/views/reponse.py

```diff
@@ -0,0 +1,53 @@
+from pyramid.httpexceptions import HTTPBadRequest, HTTPFound, HTTPNotFound
+from pyramid.request import Request
+from pyramid.response import Response
+from pyramid.view import view_config, view_defaults
+
+from zam_repondeur.models import (
+    DBSession,
+    Amendement as AmendementModel,
+    CHAMBRES,
+    AVIS,
+)
+
+
+@view_defaults(route_name="reponse_edit", renderer="reponse_edit.html")
+class ReponseEdit:
+    def __init__(self, request: Request) -> None:
+        self.request = request
+
+        chambre = request.matchdict["chambre"]
+        session = request.matchdict["session"]
+        num_texte = request.matchdict["num_texte"]
+        num = request.matchdict["num"]
+        if chambre not in CHAMBRES:
+            raise HTTPBadRequest
+        self.amendement = (
+            DBSession.query(AmendementModel)
+            .filter(
+                AmendementModel.chambre == chambre,
+                AmendementModel.session == session,
+                AmendementModel.num_texte == num_texte,
+                AmendementModel.num == num,
+            )
+            .first()
+        )
+        if self.amendement is None:
+            raise HTTPNotFound
+
+    @view_config(request_method="GET")
+    def get(self) -> dict:
+        return {"amendement": self.amendement, "avis": AVIS}
+
+    @view_config(request_method="POST")
+    def post(self) -> Response:
+        self.amendement.avis = self.request.POST["avis"]
+        self.amendement.observations = self.request.POST["observations"]
```

Comment:
Just curious, don't you need to save explicitly your modifications at some point? Magic?

Reply:
Excellent question, sir.

There is actually some kind of "magic" happening, and you can see it in these two places :

- First, in `__init__.py` ([link](https://github.com/betagouv/zam/blob/repondeur-db/repondeur/zam_repondeur/__init__.py#L15)), when initializing the app, we add [pyramid_tm](https://docs.pylonsproject.org/projects/pyramid-tm/en/latest/), which is an extension connecting the Pyramid request/response cycle with the [transaction](https://pypi.org/project/transaction/) package, which is itself a generic *transaction manager*.

- Second, in `models/base.py` ([link](https://github.com/betagouv/zam/blob/repondeur-db/repondeur/zam_repondeur/models/base.py#L6)), we also add an extension to the SQLAlchemy session factory, so the it is attached to the transaction framework.

The way it works then, is that a new transaction is started for each request, and if that request succeeds, then the transaction will be committed, and the changes will be saved to the DB ; if the request fails, though (i.e. an exception is raised), the transaction will be rolled back, and the changes will be canceled.

This is different from Django where you would do an explicit `save()`, or from the way SQLAlchemy is typically used with Flask, where you would do a explicit `db_session.commit()`.

This pattern for transaction management is indeed one of the reasons I like Pyramid a lot.